clang/lib/StaticAnalyzer/Checkers/StdVariantChecker.cpp

```diff
@@ -0,0 +1,312 @@
+  // of the first type of the possible types
+  void handleDefaultConstructor(const CallEvent &Call,
+                                CheckerContext &C) const {
+
+    const auto *AsConstructorCall = dyn_cast<CXXConstructorCall>(&Call);
+    assert(AsConstructorCall && "A constructor call must be passed!");
+
+    SVal ThisSVal = AsConstructorCall->getCXXThisVal();
+
+    const auto *const ThisMemRegion = ThisSVal.getAsRegion();
+    if (!ThisMemRegion) {
+      return;
+    }
+
+    QualType DefaultType =
+        getNthTemplateTypeArgFromVariant(ThisSVal.getType(C.getASTContext())
+                                             .getTypePtr()
+                                             ->getPointeeType()
+                                             .getTypePtr(),
+                                         0);
+
+    ProgramStateRef State = Call.getState();
+    State = State->set<VariantHeldTypeMap>(ThisMemRegion, DefaultType);
+    C.addTransition(State);
+  }
+
+  bool handleStdGetCall(const CallEvent &Call, CheckerContext &C) const {
+    ProgramStateRef State = Call.getState();
+
+    const auto &ArgType = Call.getArgSVal(0)
+                              .getType(C.getASTContext())
+                              .getTypePtr()
+                              ->getPointeeType()
+                              .getTypePtr();
+    // We have to make sure that the argument is an std::variant.
+    // There is another std::get with std::pair argument
+    if (!isStdVariant(ArgType)) {
+      return false;
+    }
+
+    // Get the mem region of the argument std::variant and get what type
+    // information is known about it.
+    const MemRegion *ArgMemRegion = Call.getArgSVal(0).getAsRegion();
+    const QualType *StoredType = State->get<VariantHeldTypeMap>(ArgMemRegion);
+    if (!StoredType) {
+      return false;
+    }
+
+    const auto &TypeOut = getFirstTemplateArgument(Call);
+    // std::get's first template parameter can be the type we want to get
+    // out of the std::variant or a natural number which is the position of
+    // the wished type in the argument std::variant's type list.
+    auto RetrievedType = [&]() {
+      switch (TypeOut.getKind()) {
+      case TemplateArgument::ArgKind::Type:
+        return TypeOut.getAsType();
+      case TemplateArgument::ArgKind::Integral:
+        // In the natural number case we look up which type corresponds to the
+        // number.
+        return getNthTemplateTypeArgFromVariant(
+            ArgType, TypeOut.getAsIntegral().getSExtValue());
+      default:
+        llvm_unreachable("An std::get's first template argument can only be a "
+                         "type or an integral");
+      }
+    }();
+
+    QualType RetrievedCanonicalType = RetrievedType.getCanonicalType();
+    QualType StoredCanonicalType = StoredType->getCanonicalType();
+    if (RetrievedCanonicalType == StoredCanonicalType) {
+      return true;
+    }
```

Comment:
Shouldn't we peel off CV quals?
To me retrieving as `const T` is the same as `T` in this context for any `T`.

Reply:
If we have
`std::variant<int, char> v;std::get<const int>(v);`

Reply:
Hm, I thought canonical means that it desugars the types aka. removes typedefs.
And a const qualified canonical is different from a non CV qualified canonical.